src/maps/mod.rs

```diff
@@ -42,32 +49,20 @@ pub struct PerfEventLostSamples {
 #[repr(C)]
 pub struct PerfEventSample {
     header: PerfEventHeader,
-    pub size: u32,
-    pub data: Vec<u8>,
+    size: u32,
+    // array to data of len `size` stored as as char[1] because Rust's
+    // DST and C's DST are are not FFI compatible. This needs to be a
+    // char[] to avoid padding issues since chars are special in c
+    // padding (in that they do not get pre-padded)
+    data: [std::os::raw::c_char; 1],
```

Comment:
The previous `Vec<u8>` implementation wasn't actually using this struct for FFI. It was instead grabbing the entire event, copying it into a `Vec<u8>` temporary buffer and then draining out the header/length by carefully offsetting based on the size of those fields.

This new implementation doesn't have to do those bit offsetting and copying because we can rely on the c representation of this struct. We can access the "data" and "size" fields directly and let the compiler take care of the offsets. The only caveat is that the `data` in `c` is a `char[]` vs the data in this `Rust` struct is `char[1]` because Rust doesn't have the concept of a dynamic array. Rust `Vec`'s are different in that they are `length` and `*T` which is not the same as `length` and `T[]`.

Reply:
Thanks!